packages/firestore/src/remote/remote_store.ts

```diff
@@ -228,26 +233,29 @@ export class RemoteStore {
         this.startWatchStream();
       }
 
-      this.setOnlineStateToUnknown();
+      this.updateOnlineState(OnlineState.Unknown);
 
       return this.fillWritePipeline(); // This may start the writeStream.
     });
   }
 
   /**
    * Temporarily disables the network. The network can be re-enabled using
-   * enableNetwork(). By default this will also update (and broadcast) our
-   * onlineState to OnlineState.Failed. If this should not happen (e.g. because
-   * this is being called as part of shutdown logic or because we're going to
-   * immediately re-enable the network), pass `true` for
-   * suppressOnlineStateFailure.
+   * enableNetwork().
    */
-  disableNetwork(suppressOnlineStateFailure?: boolean): Promise<void> {
-    if (suppressOnlineStateFailure) {
-      this.setOnlineStateToUnknown();
-    } else {
-      this.updateAndBroadcastOnlineState(OnlineState.Failed);
-    }
+  disableNetwork(): Promise<void> {
+    // Set the OnlineState to failed so get()'s return from cache, etc.
+    return this.disableNetworkInternal(OnlineState.Failed);
+  }
+
+  /**
+   * Disables the network, setting the OnlineState to the specified
+   * targetOnlineState.
+   */
+  private disableNetworkInternal(
+    targetOnlineState: OnlineState
+  ): Promise<void> {
+    this.updateOnlineState(targetOnlineState);
```

Comment:
This makes me uncomfortable because the online state handler gets called essentially before any of the actual work to disable the network has completed. While I don't think it currently does this, that handler could call back into the remote store and it would be inconsistent.

My advice would be do to make this update last.

Alternatively (but less attractively) force the update onto the async queue so that it can't run until this function returns.

Note that if we ever add a second state that can broadcast this pattern of update and broadcast if necessary stops working. It could also be worth biting the bullet and splitting the state change from the notification. I'm not sure this is really needed though and YAGNI so probably not worth it.

Reply:
Good point.  Moved the update to be last.